File: src/frontend/scripts/automation/checkout_controller.py
"""
Checkout Automation Controller

This module provides the API routes and controller logic for the AI-powered
checkout automation system. It interfaces between the Flask web application
and the Node.js automation script that uses Puppeteer and Google Gemini.
"""

import os
import json
import time
import uuid
import subprocess
import threading
import logging
from datetime import datetime
from flask import Blueprint, request, jsonify, render_template, current_app

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    handlers=[
        logging.FileHandler("automation.log"),
        logging.StreamHandler()
    ]
)
logger = logging.getLogger("checkout_automation")

# Create Blueprint for automation routes
router = Blueprint('checkout_automation', __name__)

# Dictionary to store job status information
automation_jobs = {}

def get_automation_script_path():
    """Return the path to the checkout automation script"""
    # Get the directory where this controller lives
    current_dir = os.path.dirname(os.path.abspath(__file__))
    
    # The checkout_automation.js file should be in the same directory
    return os.path.join(current_dir, 'checkout_automation.js')

def run_automation(job_id, checkout_url, product_info):
    """
    Run the checkout automation script as a subprocess
    
    Args:
        job_id (str): Unique identifier for this automation job
        checkout_url (str): URL of the checkout page to automate
        product_info (dict): Information about the product being purchased
    """
    try:
        # Update job status
        automation_jobs[job_id]['status'] = 'running'
        automation_jobs[job_id]['progress'] = 5
        automation_jobs[job_id]['logs'].append('Starting checkout automation...')
        
        # Get the path to the automation script
        script_path = get_automation_script_path()
        
        # Ensure the screenshots directory exists
        screenshots_dir = os.path.join(os.path.dirname(script_path), 'screenshots')
        os.makedirs(screenshots_dir, exist_ok=True)
        
        # Create a temporary file with the job configuration
        config_file = os.path.join(os.path.dirname(script_path), f'job_{job_id}.json')
        with open(config_file, 'w') as f:
            json.dump({
                'job_id': job_id,
                'checkout_url': checkout_url,
                'product_info': product_info,
                'timestamp': datetime.now().isoformat()
            }, f)
        
        # Log the command we're about to run
        cmd = [
            'node', 
            script_path, 
            '--checkout-url', checkout_url,
            '--job-id', job_id
        ]
        logger.info(f"Running command: {' '.join(cmd)}")
        
        # For demo purposes, we'll simulate the automation steps
        # In a real implementation, we would run the node script:
        # process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        
        # Simulate automation steps
        simulate_automation_steps(job_id)
        
        # Clean up the config file
        if os.path.exists(config_file):
            os.remove(config_file)
            
    except Exception as e:
        logger.error(f"Error running automation: {str(e)}")
        automation_jobs[job_id]['status'] = 'error'
        automation_jobs[job_id]['error'] = str(e)
        automation_jobs[job_id]['logs'].append(f'Error: {str(e)}')

def simulate_automation_steps(job_id):
    """
    Simulate the automation steps for demo purposes
    In a real implementation, this would be handled by the Node.js script
    """
    def update_progress():
        try:
            # Step 1: Analyzing the checkout page
            automation_jobs[job_id]['progress'] = 10
            automation_jobs[job_id]['logs'].append('Analyzing checkout page...')
            time.sleep(2)
            
            # Step 2: Filling out shipping information
            automation_jobs[job_id]['progress'] = 30
            automation_jobs[job_id]['logs'].append('Filling out shipping information...')
            time.sleep(2)
            
            # Step 3: Filling out payment information
            automation_jobs[job_id]['progress'] = 60
            automation_jobs[job_id]['logs'].append('Entering payment details...')
            time.sleep(2)
            
            # Step 4: Confirming order
            automation_jobs[job_id]['progress'] = 85
            automation_jobs[job_id]['logs'].append('Confirming order...')
            time.sleep(2)
            
            # Step 5: Order completed
            automation_jobs[job_id]['progress'] = 100
            automation_jobs[job_id]['status'] = 'completed'
            automation_jobs[job_id]['logs'].append('Order completed successfully!')
            
            # Generate fake order number
            order_number = f"AI-{uuid.uuid4().hex[:8].upper()}"
            automation_jobs[job_id]['order_number'] = order_number
            automation_jobs[job_id]['logs'].append(f'Order number: {order_number}')
            
            logger.info(f"Automation job {job_id} completed successfully")
        except Exception as e:
            logger.error(f"Error in simulation thread: {str(e)}")
            automation_jobs[job_id]['status'] = 'error'
            automation_jobs[job_id]['error'] = str(e)
    
    # Run the simulation in a separate thread
    thread = threading.Thread(target=update_progress)
    thread.daemon = True
    thread.start()

@router.route('/checkout', methods=['POST'])
def start_checkout_automation():
    """API endpoint to start the checkout automation process"""
    try:
        # Get the checkout URL and product info from the request
        data = request.json
        checkout_url = data.get('checkout_url')
        product_info = data.get('product_info', {})
        
        if not checkout_url:
            return jsonify({
                'success': False,
                'error': 'Checkout URL is required'
            }), 400
        
        # Generate a unique job ID
        job_id = str(uuid.uuid4())
        
        # Initialize job status
        automation_jobs[job_id] = {
            'id': job_id,
            'status': 'initializing',
            'progress': 0,
            'created_at': datetime.now().isoformat(),
            'checkout_url': checkout_url,
            'product_info': product_info,
            'logs': ['Initializing checkout automation...']
        }
        
        # Start the automation process in a separate thread
        thread = threading.Thread(
            target=run_automation, 
            args=(job_id, checkout_url, product_info)
        )
        thread.daemon = True
        thread.start()
        
        logger.info(f"Started automation job {job_id} for URL: {checkout_url}")
        
        return jsonify({
            'success': True,
            'job_id': job_id,
            'message': 'Checkout automation started'
        })
        
    except Exception as e:
        logger.error(f"Error starting automation: {str(e)}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@router.route('/status/<job_id>', methods=['GET'])
def get_automation_status(job_id):
    """API endpoint to get the status of an automation job"""
    try:
        if job_id not in automation_jobs:
            return jsonify({
                'success': False,
                'error': 'Job not found'
            }), 404
        
        job = automation_jobs[job_id]
        
        # Clean up logs, filter out None values
        logs = [log for log in job.get('logs', []) if log is not None]
        
        return jsonify({
            'success': True,
            'status': job.get('status', 'unknown'),
            'progress': job.get('progress', 0),
            'logs': logs,
            'error': job.get('error'),
            'order_number': job.get('order_number')
        })
        
    except Exception as e:
        logger.error(f"Error getting job status: {str(e)}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@router.route('/logs/<job_id>', methods=['GET'])
def get_automation_logs(job_id):
    """API endpoint to get the logs for an automation job"""
    try:
        if job_id not in automation_jobs:
            return jsonify({
                'success': False,
                'error': 'Job not found'
            }), 404
        
        logs = automation_jobs[job_id].get('logs', [])
        # Clean up logs, filter out None values
        logs = [log for log in logs if log is not None]
        
        return jsonify({
            'success': True,
            'logs': logs
        })
        
    except Exception as e:
        logger.error(f"Error getting job logs: {str(e)}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

def registerTestRoutes(app):
    """Register additional test routes for the automation system"""
    
    @app.route('/test-automation')
    def test_automation_page():
        """Test page to demonstrate the checkout automation"""
        return render_template('test_automation.html')
    
    @app.route('/api/automation/test', methods=['POST'])
    def test_automation():
        """API endpoint to test the automation system"""
        try:
            # Generate a unique job ID
            job_id = str(uuid.uuid4())
            
            # Initialize job status
            automation_jobs[job_id] = {
                'id': job_id,
                'status': 'initializing',
                'progress': 0,
                'created_at': datetime.now().isoformat(),
                'checkout_url': 'http://example.com/test-checkout',
                'product_info': {'title': 'Test Product', 'price': '₹499'},
                'logs': ['Initializing test automation...']
            }
            
            # Start the simulation in a separate thread
            thread = threading.Thread(target=simulate_automation_steps, args=(job_id,))
            thread.daemon = True
            thread.start()
            
            logger.info(f"Started test automation job {job_id}")
            
            return jsonify({
                'success': True,
                'job_id': job_id,
                'message': 'Test automation started'
            })
            
        except Exception as e:
            logger.error(f"Error starting test automation: {str(e)}")
            return jsonify({
                'success': False,
                'error': str(e)
            }), 500 

File: src/frontend/scripts/automation/checkout_controller.js
/**
 * Checkout Controller
 * 
 * This script provides an Express.js controller to handle checkout automation requests.
 * It interfaces between the frontend UI and the Puppeteer automation backend.
 */

const express = require('express');
const router = express.Router();
const path = require('path');
const fs = require('fs');
const { automateCheckout } = require('./checkout_automation');

// Storage for tracking automation jobs
const automationJobs = new Map();

/**
 * Start a new checkout automation process
 * POST /api/automation/checkout
 */
router.post('/checkout', async (req, res) => {
  try {
    const { checkoutUrl, productInfo, customerProfile } = req.body;
    
    if (!checkoutUrl) {
      return res.status(400).json({
        success: false,
        error: 'Missing required parameter: checkoutUrl'
      });
    }
    
    // Generate a unique job ID
    const jobId = `job_${Date.now()}_${Math.random().toString(36).substring(2, 15)}`;
    
    // Store job information
    automationJobs.set(jobId, {
      status: 'pending',
      checkoutUrl,
      productInfo: productInfo || {},
      customerProfile: customerProfile || 'default',
      createdAt: new Date(),
      logs: ['Job created. Waiting to start...']
    });
    
    // Start the automation process in the background
    startAutomationJob(jobId);
    
    // Return the job ID to the client
    return res.json({
      success: true,
      jobId,
      message: 'Checkout automation job created'
    });
    
  } catch (error) {
    console.error('Error creating automation job:', error);
    return res.status(500).json({
      success: false,
      error: error.message
    });
  }
});

/**
 * Get status of a checkout automation job
 * GET /api/automation/checkout/:jobId
 */
router.get('/checkout/:jobId', (req, res) => {
  const { jobId } = req.params;
  
  if (!automationJobs.has(jobId)) {
    return res.status(404).json({
      success: false,
      error: 'Job not found'
    });
  }
  
  const job = automationJobs.get(jobId);
  
  // If the job is completed, return the full result
  if (job.status === 'completed' || job.status === 'failed') {
    return res.json({
      success: true,
      job
    });
  }
  
  // For in-progress jobs, return the status and logs
  return res.json({
    success: true,
    status: job.status,
    logs: job.logs,
    progress: job.progress || 0
  });
});

/**
 * Cancel a running checkout automation job
 * DELETE /api/automation/checkout/:jobId
 */
router.delete('/checkout/:jobId', (req, res) => {
  const { jobId } = req.params;
  
  if (!automationJobs.has(jobId)) {
    return res.status(404).json({
      success: false,
      error: 'Job not found'
    });
  }
  
  const job = automationJobs.get(jobId);
  
  // If the job is already completed or failed, just return the status
  if (job.status === 'completed' || job.status === 'failed') {
    return res.json({
      success: true,
      message: `Job already ${job.status}`,
      job
    });
  }
  
  // Mark the job as cancelled
  job.status = 'cancelled';
  job.logs.push('Job cancelled by user');
  
  // Try to terminate the browser process if it exists
  if (job.browser) {
    try {
      job.browser.close().catch(() => {});
    } catch (error) {
      console.warn('Error closing browser:', error);
    }
  }
  
  return res.json({
    success: true,
    message: 'Job cancelled successfully'
  });
});

/**
 * Start an automation job in the background
 * @param {string} jobId - The ID of the job to start
 */
async function startAutomationJob(jobId) {
  const job = automationJobs.get(jobId);
  
  if (!job) {
    console.error(`Job ${jobId} not found`);
    return;
  }
  
  try {
    // Update job status
    job.status = 'running';
    job.logs.push('Starting checkout automation...');
    job.startedAt = new Date();
    
    // Run the automation
    const result = await automateCheckout(
      job.checkoutUrl,
      job.productInfo,
      job.customerProfile
    );
    
    // Update job with result
    job.status = result.success ? 'completed' : 'failed';
    job.logs.push(result.success 
      ? `Checkout completed successfully. Order number: ${result.orderNumber}` 
      : `Checkout failed: ${result.error}`);
    job.result = result;
    job.completedAt = new Date();
    
    // If screenshot was taken, save the path
    if (result.screenshots) {
      job.screenshots = result.screenshots;
    }
    
  } catch (error) {
    console.error(`Error in automation job ${jobId}:`, error);
    
    // Update job with error
    job.status = 'failed';
    job.logs.push(`Error: ${error.message}`);
    job.error = error.message;
    job.completedAt = new Date();
  }
  
  // Clean up older jobs (keep for 1 hour)
  cleanupOldJobs();
}

/**
 * Clean up completed jobs older than 1 hour
 */
function cleanupOldJobs() {
  const oneHourAgo = new Date(Date.now() - 60 * 60 * 1000);
  
  for (const [jobId, job] of automationJobs.entries()) {
    if ((job.status === 'completed' || job.status === 'failed' || job.status === 'cancelled') &&
        job.completedAt && job.completedAt < oneHourAgo) {
      automationJobs.delete(jobId);
    }
  }
}

/**
 * Register additional routes for testing/demo purposes
 */
function registerTestRoutes(app) {
  // Serve screenshots
  app.get('/api/automation/screenshots/:filename', (req, res) => {
    const { filename } = req.params;
    const screenshotPath = path.join(process.cwd(), filename);
    
    if (fs.existsSync(screenshotPath)) {
      res.sendFile(screenshotPath);
    } else {
      res.status(404).send('Screenshot not found');
    }
  });
  
  // List all jobs (for debugging/admin)
  app.get('/api/automation/jobs', (req, res) => {
    const jobs = Array.from(automationJobs.entries()).map(([id, job]) => ({
      id,
      status: job.status,
      product: job.productInfo?.title || 'Unknown product',
      created: job.createdAt,
      completed: job.completedAt
    }));
    
    res.json({ jobs });
  });
}

module.exports = {
  router,
  registerTestRoutes,
  startAutomationJob
}; 